flow/directives.py

```diff
@@ -0,0 +1,328 @@
+
+    def __str__(self):
+        return str(dict(self))
+
+    def __repr__(self):
+        return "Directives({})".format(str(self))
+
+    def update(self, other, aggregate=False, job=None, parallel=False):
+        if aggregate:
+            self._aggregate(other, job=job, parallel=parallel)
+        else:
+            super().update(other)
+
+    def evaluate(self, job):
+        for key, value in self.items():
+            self[key] = _evaluate(value, job)
+
+    def _aggregate(self, other, job=None, parallel=False):
+        self.evaluate(job)
+        other.evaluate(job)
+        agg_func_attr = "parallel" if parallel else "serial"
+        for name in self._defined_directives:
+            agg_func = getattr(self._directive_definitions[name], agg_func_attr)
+            default_value = self._directive_definitions[name].default
+            other_directive = other.get(name, default_value)
+            directive = self[name]
+            if other_directive is None:
+                continue
+            else:
+                self._defined_directives[name] = agg_func(directive,
+                                                          other_directive)
+
+
+def _evaluate(value, job=None):
+    if callable(value):
+        if job is None:
+            raise RuntimeError("job must be specified when evaluating a callable directive.")
+        else:
+            return value(job)
+    else:
+        return value
+
+
+class _OnlyType:
+    def __init__(self, type_, preprocess=None, postprocess=None):
+        def identity(v):
+            return v
+
+        self.type = type_
+        self.preprocess = identity if preprocess is None else preprocess
+        self.postprocess = identity if postprocess is None else postprocess
+
+    def __call__(self, v):
+        return self.postprocess(self._validate(self.preprocess(v)))
+
+    def _validate(self, v):
+        if isinstance(v, self.type):
+            return v
+        else:
+            try:
+                return self.type(v)
+            except Exception:
+                raise TypeError("Expected an object of type {}. Received {} "
+                                "of type {}".format(self.type, v, type(v)))
+
+
+def _raise_below(value):
+    def is_greater(v):
+        try:
+            if v < value:
+                raise ValueError
+        except (TypeError, ValueError):
+            raise ValueError("Expected a number greater than {}. Received {}"
+                             "".format(value, v))
+        return v
+    return is_greater
+
+
+_NP_DEFAULT = 1
+
+
+def _finalize_np(np, directives):
+    """Return the actual number of processes/threads to use.
+
+    We check the default np because when aggregation occurs we multiple the
+    number of MPI ranks and OMP_NUM_THREADS. If we always took the greater of
+    the given NP and ranks * threads then after aggregating we will inflate the
+    number of processors needed as (r1 * t1) + (r2 * t2) <= (r1 * r2 * t1 * t2)
```

Comment:
I'm not sure how you got `r1 * r2 * t1 * t2` in the original expression... is this supposed to say:
```suggestion
    number of processors needed as (r1 * t1) + (r2 * t2) <= (r1 + r2) * (t1 + t2)
```
But also my suggested expression isn't always true.